### ISSUE TITLE
Add "modes" to compiletest, for running all tests with NLL enabled and comparing with master

### ISSUE THREAD

**Issue by pnkfelix** [rust-lang/rust#48879]
The NLL team has been trying to find a good workflow for evaluating the discrepancies between AST borrowck and NLL-based borrowck.

 * Past workflows have developed and used things like `-Z borrowck=compare` and the https://github.com/pnkfelix/nll-probe tool, but here I'm going to try to avoid a deep dive into how those have failed to satisfy our needs.

Our goal here: We want a way to track the current state of NLL, including the cases where there are known deviations, in a manner that allows immediate analysis of that state to determine things like:
 * "how many test cases does NLL deviate from AST borrowck",
 * "what do the deviations look like", or
 * "does this deviation represent a bug that needs to be fix? Or is it an improvement on the AST borrowck?"

Here's a new proposal for a (hopefully) relatively small change to `compiletest` that should yield an easier workflow to answer questions like those above, at least for the `ui/` subset of our tests.

 1. Context: The `ui/` tests are set up so that each test consists of an input `$file.rs`, and a set of expected outputs in `$file.stderr` (compilation error messages) and `$file.stdout` (non erroneous compiler output; rarely used). (For more info, see [this chapter in the rustc-guide](https://rust-lang-nursery.github.io/rustc-guide/tests/adding.html#guide-to-the-ui-tests).)
 1. Add a "mode" argument to `compiletest`, encoded either as a command line parameter or as an environment variable, or both. (The first mode we'll support will be "nll", which tells `compiletest` to pass `-Z nll` in addition to any other flags when invoking `rustc`, at least for the `ui/` tests).
 2. When running under a particular mode, if there is `$file.$mode.stderr` file, then this file will be used as the source of "acceptable output". If there is no such file, then `compiletest` will fallback to the regular filename `$file.stderr`.
 3. One complication: UI tests also include "inline" comments of the form `//~ ERROR` that indicate what error is expected on which line (these are mildly redundant with the stderr files above). Because messages may differ in the mode M, but we don't want to edit the sources too much, compiletest should just ignore the `//~ ERROR` annotations when running in a particular mode. We will still see the errors that occur from the stderr output, it's just less convenient.
 4. To ensure that we are tracking discrepancies somewhere, whenever there is a `$file.$mode.stderr`, then some tool (probably `compiletest`, but maybe `tidy`?) will be responsible to checking that `$file.rs` somewhere contains a comment that explains the source of the discrepancy. This could be a specially formatted FIXME, if the new behavior seems worse than before:
```rust
// $mode FIXME(#123) -- summary of discrepancy caused by NLL bug with corresponding issue num
```
or a "YAYME" comment for when the new behavior is an improvement:
```rust
// $mode YAYME(#123) -- summary of a beneficial discrepancy
```
(presumably the ticket linked by YAYME would just be the tracking issue for NLL or whatever other mode is being tested).

Benefits of the proposed system:

 * To find the cases that have discrepancies for nll, one can use `ls *.nll.stderr`
 * To find what a given discrepancy looks like, one can use `diff onetest.stderr onetest.nll.stderr`
 * To see if a discrepancy is a bug or not, grep for FIXME or YAYME in the .rs files.
   * This workflow is perhaps not ideal; @nikomatsakis has pointed out that it might be nicer if these comments somehow lived in the `*.$mode.stderr` files.

# Open Questions (to be resolved by implementor)

 * [ ] What should `compiletest` do about occurrences of `//~ ERROR` in the source text? In particular, should it check that the error output still has those cases, even when running under a given mode?
   * The current inclination of @pnkfelix and @nikomatsakis is that it is actually *okay* for `compiletest` to *ignore* the `//~ ERROR` annotations when running under a given mode. The reasoning here is this: the `//~ ERROR` annotations will already get checked by compiletest runs that *don't* have a mode. We probably *don't* want to force an error when there's a discrepancy when running under a given mode; any discrepancies, including any of those errors disappearing, should be accounted for in the linked `// $mode FIXME/YAYME`  issue, and we want to allow them to disappear or differ.

**Comment by nikomatsakis** [comment]
I'm marking this as belonging to both @rust-lang/wg-compiler-nll and @rust-lang/wg-traits -- this is because we are both going to need this tool! In general, having a mode like this will be useful anytime that we plan to make major alterations to some core component of the system.

**Comment by nikomatsakis** [comment]
### Mentoring instructions

I'm going to assume we will use a command line option to drive this new mode. Let's call "compare-mode". Then we can run the tests with `./x.py test foo --test-args --compare-mode nll` or whatever. The compile test harness as a configuration that is loaded here:

https://github.com/rust-lang/rust/blob/fedce67cd21dc08ece5a484fe1a060346acac98a/src/tools/compiletest/src/main.rs#L58

So we have to extend `parse_config` with some new option `--compare-mode` or whatever that takes a string.

https://github.com/rust-lang/rust/blob/fedce67cd21dc08ece5a484fe1a060346acac98a/src/tools/compiletest/src/main.rs#L68

That comparison mode will be translated to some arguments to rustc. Those arguments are arranged in this function:

https://github.com/rust-lang/rust/blob/fedce67cd21dc08ece5a484fe1a060346acac98a/src/tools/compiletest/src/runtest.rs#L1584

Probably it's best we allow a mapping between the comparison mode and the arguments (that mapping can just be hardcoded into this function though). For example, NLL at least currently really wants three flags added:

```
match self.config.comparison_mode {
    Some(ComparisonMode::NLL) => {
        rustc.args(&["-Znll", "-Zborrowck=mir", "-Ztwo-phase-borrows"]);
    }
    None => { }
}
```

This is where the code currently finds the expected files. We'll want to alter these functions to take the comparison mode into account. That's probably a new option to be added, along with revision.

https://github.com/rust-lang/rust/blob/fedce67cd21dc08ece5a484fe1a060346acac98a/src/tools/compiletest/src/runtest.rs#L2504-L2508

Right now, if the revision is `Some`, then we look for `$test.$rev.stderr`, I think we want to look for `$test.$rev.$mode.stderr` in that case, so that we can always do `ls *.nll.stderr` to get all things with a given mode. (Or maybe we should make it `.mode_$mode.stderr` or something more distinctive?)

Anyway, we'll have to alter this to search for the expected output with the given mode and -- if not found -- to then try with a mode of `None`.

Finally, to skip checking for `//~ ERROR` comments, if a comparison mode is set, we just don't do any of this stuff:

https://github.com/rust-lang/rust/blob/fedce67cd21dc08ece5a484fe1a060346acac98a/src/tools/compiletest/src/runtest.rs#L2548-L2554

Uuuuh at that point we are basically done with the compiletest changes, right?

Then the last step is to alter tidy to enforce the FIXME convention, but I don't know much about that. We'll figure that out when we get there.

**Comment by kennytm** [comment]
In compiletest there is already the concept of revisions which y'all are already using in [compile-fail](https://github.com/rust-lang/rust/blob/master/src/test/compile-fail/assign-imm-local-twice.rs) [tests](https://github.com/rust-lang/rust/blob/master/src/test/compile-fail/regions-struct-not-wf.rs). We just need to extend it to UI tests.

**Comment by pnkfelix** [comment]
@kennytm this is different interface from revisions. Revisions are encoded with flags on a test by test basis. A mode is given more privileged status within `compiletest`. We explicitly do not want to add a revision for every test to represent running with and without NLL.

(There's some chance that code might be shared within `compiletest` between the support for revisions and that for modes, but in terms of UI, the interface is different here.)

**Comment by nikomatsakis** [comment]
Also, @kennytm, I opened this issue regarding extending revisions to UI tests (https://github.com/rust-lang/rust/issues/48878). That said, I think they already work, and we're just not using them. I've got to do some local tests.

**Comment by nikomatsakis** [comment]
I think though we *will* eventually want to specify the intersection of the "comparison mode" with a revision. There are a lot of tests that are already encoded with `lexical` and `nll` modes, for example -- in those cases, if we have a revision named `nll` and we are in comparison mode `nll`, it'd be sort of nice to just ignore the test, or do *something* smart. But I guess we can leave that for later.

**Comment by memoryleak47** [comment]
I'd like to try this one. :)

**Comment by nikomatsakis** [comment]
@memoryleak47 that would be super duper. @pnkfelix and I were just saying that this remains pretty high priority. Let me know if I can help in any way. (Of course there are already [mentoring instructions](https://github.com/rust-lang/rust/issues/48879#issuecomment-371828898).)

**Comment by pnkfelix** [comment]
I'm going to actually reopen this issue as a way to track when we've gotten this compare-mode into a state where we actually have files checked in that we are referencing against.

(I just discovered that tidy complains about a `augemented-assignments.nll.stderr` file that I had lying around, so we clearly cannot yet reach that point without some more work.)

It might be good to even keep this issue open until we have bors set up to be running `compare-more=nll` against the ui tests on some dedicated target like linux.

**Comment by pnkfelix** [comment]
Once rust-lang/rust#49900 lands, this issue will be almost completely resolved; I think the only task I could imagine beyond that would be to expand rust-lang/rust#49900 (and also `compare-mode`?) to work on all the tests, not just the ones in `src/test/ui`.

**Comment by nikomatsakis** [comment]
I think we can call this done for now until we have a concrete extension in mind.

**Comment by pnkfelix** [comment]
namely @nikomatsakis and @pnkfelix agreed that any remaining work here is best spent on resolving #44844  (*plus* porting the `run-pass/` tests to `ui/`, which should probably be filed as another issue).

**Comment by nikomatsakis** [comment]
In particular closed in favor of https://github.com/rust-lang/rust/issues/44844
